source/fundamentals/typescript.txt

```diff
@@ -0,0 +1,160 @@
+
+Extend Document
+~~~~~~~~~~~~~~~
+
+The following classes accept any type that extends the ``Document``
+interface:
+
+- :node-api-4.0:`Collection <classes/collection.html>`
+- :node-api-4.0:`ChangeStream <classes/changestream.html>`
+
+You can pass a type parameter that extends the ``Document`` interface like this:
+
+.. literalinclude:: /code-snippets/typescript/extend-document.ts
+   :language: typescript
+   :linenos:
+
+.. important:: Keys Not in Type Parameter Receive ``any`` Type
+
+   Keys not listed in your specified type parameter receive the ``any`` type.
+   The following code snippet demonstrates this behavior:
+
+   .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+      :language: typescript
+      :linenos:
+      :start-after: start-no-key
+      :end-before: end-no-key
+
+Any Type
+~~~~~~~~
+
+The following classes accept any type parameter:
+
+- :node-api-4.0:`FindCursor <classes/findcursor.html>`
+- :node-api-4.0:`AggregationCursor <classes/aggregationcursor.html>`
+
+You can find a code snippet that shows how to specify a type for the ``FindCursor``
+class in the
+:ref:`Find Multiple Documents Usage Example <node-driver-find-usage-example-code-snippet>`.
+
+Limitations
+-----------
+
+.. _node-driver-typescript-limitations-dot-notation:
+
+The driver cannot infer the type of values with keys containing **dot
+notation**. Dot notation is a property access syntax for navigating BSON objects.
+Click on the tabs to see code snippets that highlight this behavior:
+
+.. tabs::
+
+  .. tab:: Dot Notation
+     :tabid: dot-notation
+  
+     The following code snippet does not raise a type error:
+
+     .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+        :language: typescript
+        :linenos:
+        :start-after: start-no-error
+        :end-before: end-no-error
+
+  .. tab:: Nested Objects
+     :tabid: nested-objects
+    
+     The following code snippet raises a type error:
+
+     .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+        :language: typescript
+        :linenos:
+        :start-after: start-error
+        :end-before: end-error
+
+     This is the error:
+
+     .. code-block:: text
+
+        Type 'string' is not assignable to type 'number'.
+
+Despite the lack of type safety, we still recommend that you use dot notation to
+access nested fields in query and update documents when you use TypeScript. You
+must manually check that your nested field values have your intended type.
+
+.. note:: Reason To Use Dot Notation
+
+   In the MongoDB Query Language, you must match a subdocument exactly
+   when specifying subdocuments in a query. Dot notation allows you to query
+   nested fields without matching subdocuments exactly.  
+
+   To show this behavior, lets say you have a collection containing
+   only the following document:
+
+   .. code-block:: json
+
+      {field: {s1: "hi", s2: "bye"}}
```

Comment:
let's use consistent quotes and curly brace spacing in the inlined examples here

Reply:
Ran through formatter!